src/main/java/io/leonis/subra/math/filter/MovingBallsKalmanFilter.java

```diff
@@ -0,0 +1,103 @@
+package io.leonis.subra.math.filter;
+
+import io.leonis.algieba.filter.KalmanFilter;
+import io.leonis.algieba.statistic.SimpleDistribution;
+import io.leonis.subra.game.data.*;
+import io.leonis.zosma.game.engine.Deducer;
+import org.nd4j.linalg.api.ndarray.INDArray;
+import org.nd4j.linalg.factory.Nd4j;
+import org.reactivestreams.Publisher;
+import reactor.core.publisher.Flux;
+
+/**
+ * The Class MovingBallsKalmanFilter.
+ *
+ * @author Rimon Oz
+ */
+public class MovingBallsKalmanFilter<I extends MovingBall.SetSupplier & Referee.Supplier>
+    implements Deducer<I, MovingBall> {
+
+  private final KalmanFilter kalmanFilter = new KalmanFilter();
+
+  public static final INDArray MEASUREMENT_TRANSITION_MATRIX = Nd4j.create(
+      new double[]{
+          1, 0, 0, 0, 0, 0, 0,
+          0, 1, 0, 0, 0, 0, 0,
+          0, 0, 1, 0, 0, 0, 0,
+          0, 0, 0, 1, 0, 0, 0,
+          0, 0, 0, 0, 1, 0, 0,
+          0, 0, 0, 0, 0, 1, 0,
+          0, 0, 0, 0, 0, 0, 1
+      },
+      new int[]{7, 7});
+  public static final INDArray MEASUREMENT_COVARIANCE_MATRIX = Nd4j.create(
+      new double[]{
+          0.01f, 0, 0, 0, 0, 0, 0,
+          0, 0.01f, 0, 0, 0, 0, 0,
+          0, 0, 0.01f, 0, 0, 0, 0,
+          0, 0, 0, 0.01f, 0, 0, 0,
+          0, 0, 0, 0, 0.01f, 0, 0,
+          0, 0, 0, 0, 0, 0.01f, 0,
+          0, 0, 0, 0, 0, 0, 0.01f
+      },
+      new int[]{7, 7});
+  public static final INDArray CONTROL_TRANSITION_MATRIX = Nd4j.create(
+      new double[]{
+          1, 0, 0, 0, 0, 0, 0,
+          0, 1, 0, 0, 0, 0, 0,
+          0, 0, 1, 0, 0, 0, 0,
+          0, 0, 0, 1, 0, 0, 0,
+          0, 0, 0, 0, 1, 0, 0,
+          0, 0, 0, 0, 0, 1, 0,
+          0, 0, 0, 0, 0, 0, 1
+      },
+      new int[]{7, 7});
+  public static final INDArray PROCESS_COVARIANCE_MATRIX = Nd4j.create(
+      new double[]{
+          0.01f, 0, 0, 0, 0, 0, 0,
+          0, 0.01f, 0, 0, 0, 0, 0,
+          0, 0, 0.01f, 0, 0, 0, 0,
+          0, 0, 0, 0.01f, 0, 0, 0,
+          0, 0, 0, 0, 0.01f, 0, 0,
+          0, 0, 0, 0, 0, 0.01f, 0,
+          0, 0, 0, 0, 0, 0, 0.01f
+      },
+      new int[]{7, 7});
+
+  @Override
+  public Publisher<MovingBall> apply(final Publisher<I> inputPublisher) {
+    return Flux.from(inputPublisher)
+        .scan(
+            new MovingBall.State(0, 0, 0, 0, 0, 0, 0),
+            (previousResult, input) -> input.getBalls().stream()
+                .reduce((previousBall, foundPlayer) ->
+                    new MovingBall.State(
+                        this.kalmanFilter.apply(
+                            getStateTransitionMatrix(
+                                (input.getReferee().getTimestamp()
```

Comment:
Why the referee timestamp? Seems quite random. Shouldn't you use the ball's timestamp?